controller/src/beerocks/master/son_master_thread.cpp

```diff
@@ -369,23 +368,92 @@ bool master_thread::handle_cmdu_1905_autoconfiguration_WSC(Socket *sd,
-    if (tlvAp == nullptr) {
-        if (cmdu_rx.addClass<wfa_map::tlvApRadioIdentifier>()) {
-            return true;
-        }
-        LOG(DEBUG) << "Failed to get APRadioBasicCapabilities";
+    if (cmdu_rx.getNextTlvType() == wfa_map::TLV_AP_RADIO_IDENTIFIER)
+        return true;
+
+    LOG(DEBUG) << "Received AP_AUTOCONFIGURATION_WIFI_SIMPLE_CONFIGURATION_MESSAGE";
+    /**
+    * @brief Parse AP-Autoconfiguration WSC which should include one AP Radio Basic Capabilities
+    * TLV and one WSC TLV containing M1
+    */
+    auto ap = cmdu_rx.addClass<wfa_map::tlvApRadioBasicCapabilities>();
+    if (ap == nullptr) {
+        LOG(ERROR) << "Failed to get APRadioBasicCapabilities";
         return false;
     }
 
-    auto tlvWscM1 = cmdu_rx.addClass<ieee1905_1::tlvWscM1>();
-    if (tlvWscM1 == nullptr) {
+    // Parse M1 TLV
+    auto m1 = cmdu_rx.addClass<ieee1905_1::tlvWscM1>();
+    if (m1 == nullptr) {
         LOG(ERROR) << "Error creating tlvWscM1";
         return false;
     }
 
     //TODO autoconfig process the rest of the class
     //TODO autoconfig add support for none intel agents
     //TODO autoconfig Keep intel agent support only as intel enhancements
+    /**
+     * @brief Reply with AP-Autoconfiguration WSC with a single AP Radio Identifier TLV
+     * and one (TODO do we need more?) WSC TLV containing M2.
+     */
+    auto c = cmdu_tx.create(
+        0, ieee1905_1::eMessageType::AP_AUTOCONFIGURATION_WIFI_SIMPLE_CONFIGURATION_MESSAGE);
+    if (!c) {
+        LOG(ERROR) << "Create AP_AUTOCONFIGURATION_WIFI_SIMPLE_CONFIGURATION_MESSAGE response";
+        return false;
+    }
+    // All attributes which are not explicitely set below are set to
+    // default by the TLV factory, see WSC_Attributes.yml
+    auto ruid = cmdu_tx.addClass<wfa_map::tlvApRadioIdentifier>();
+    if (!ruid) {
+        LOG(ERROR) << "error creating tlvApRadioIdentifier TLV";
+        return false;
+    }
+    // Let the Agent keep its Radio Identifier
+    ruid->radio_uid() = ap->radio_uid();
+
+    auto m2 = cmdu_tx.addClass<ieee1905_1::tlvWscM2>();
+    if (m2 == nullptr) {
+        LOG(ERROR) << "Error creating tlvWscM1";
+        return false;
+    }
+
+    m2->M2Frame().message_type_attr.data = WSC::WSC_MSG_TYPE_M2;
+    string_utils::copy_string(m2->M2Frame().manufacturer_attr.data, "Intel",
+                              m2->M2Frame().manufacturer_attr.data_length);
+    string_utils::copy_string(m2->M2Frame().model_name_attr.data, "Ubuntu",
+                              m2->M2Frame().model_name_attr.data_length);
+    string_utils::copy_string(m2->M2Frame().model_number_attr.data, "18.04",
+                              m2->M2Frame().model_number_attr.data_length);
+    std::string manufacturer = std::string(m1->M1Frame().manufacturer_attr.data,
+                                           m1->M1Frame().manufacturer_attr.data_length);
+
+    if (!manufacturer.compare("Intel")) {
```

Comment:
what won't you do that before all the M2 logic?

Reply:
I don't follow - we read the manufacturer from the M2 message, how can it be done outside the M2 parsing logic?